### PR TITLE
feat: Expire conditions

### PR DESCRIPTION
**Pull request by cshezi**
# SPDX-License-Identifier: Apache-2.0

## What did we change?
1. We add new logic for expiring conditions
2. We test the new logic for expiring conditions
3. We validated incoming data (params and body)
4. We documented our new endpoints
5. We added 2 methods from our endpoints
6. We lowered cases of params

## Why are we doing this?
1. To allow user to expire or update conditions
2. To test code integrity or logic
3. To ensure that incoming data is correct
4. To allow ease of use for other integrators
5. To allow external HTTP request to reach our endpoints
6. Params are always in lowercase 

## How was it tested?
- [x] Locally
- [ ] Development Environment
- [ ] Not needed, changes very basic
- [x] Husky successfully run
- [x] Unit tests passing and Documentation done

image

### CODE REVIEW

**Review by Justus-at-Tazama — APPROVED** [review]

File: src/utils/condition-validation.ts [hunk 1]
```diff
@@ -42,4 +42,21 @@ const checkConditionValidity = (condition: EntityCondition | AccountCondition):
...
+    }
+  }
+
+  if (!xprtnDtTm) {
+    return new Date().toISOString();
+  }
+  return xprtnDtTm;
+};
+
 export default checkConditionValidity;
```
Comment (Lenbkr) [inline src/utils/condition-validation.ts:62]: 1. This function is trying to do too much on it's own. I suggest you break it up into easier managed pieces. 
Here's a starting snippet
```
export const convertDateToISO = (dateStr: string): string | null {
  if (!isDateValid(dateStr)) {
    throw Error('Error: Date provided was invalid.')
  }

  const date = new Date(dateStr);
  return date.toISOString();
}

const isDateValid = (dateStr: string): boolean {
  const date = new Date(dateStr);
  return !isNaN(date.getTime());
}

export const hasDateExpired = (date: Date): boolean {
  const now = new Date();
  return date < now;
}
```

2. Also you have a default export and now non default exports. I suggest you keep it consistent.

File: src/logic.service.ts [hunk 2]
```diff
@@ -306,3 +367,68 @@ export const handleGetConditionsForAccount = async (params: GetAccountConditions
...
+  if (
+    !creditorByEdge.filter((eachDocument) => eachDocument.result._id).length &&
+    !debtorByEdge.filter((eachDocument) => eachDocument.result._id).length
+  ) {
+    return { code: 404, message: 'Account does not exist in the database.' };
+  }
+
+  if (
+    creditorByEdge.filter((eachDocument) => eachDocument.condition.xprtnDtTm).length ||
+    debtorByEdge.filter((eachDocument) => eachDocument.condition.xprtnDtTm).length
```
Comment (Lenbkr) [inline src/logic.service.ts:408]: Use some rather than filter

File: src/interface/queryAccountCondition.ts [hunk 3]
```diff
@@ -1,6 +1,7 @@
 export interface GetAccountConditions {
```
Comment (Lenbkr) [inline src/interface/queryAccountCondition.ts:1]: Omit Get from the interface name

File: src/logic.service.ts (hunk 2 above)
Comment (Lenbkr) [reply src/logic.service.ts:408]: or every if you want to validate each property

File: src/logic.service.ts [hunk 4]
```diff
@@ -188,6 +188,67 @@ export const handleGetConditionsForEntity = async (params: GetEntityConditions):
...
+  if (
+    !creditorByEdge.filter((eachDocument) => eachDocument.result._id).length &&
+    !debtorByEdge.filter((eachDocument) => eachDocument.result._id).length
+  ) {
+    return { code: 404, message: 'Entity does not exist in the database.' };
+  }
+
+  if (
+    creditorByEdge.filter((eachDocument) => eachDocument.condition.xprtnDtTm).length ||
+    debtorByEdge.filter((eachDocument) => eachDocument.condition.xprtnDtTm).length
```
[the remark this answers is not in the document]
Comment (Lenbkr) [reply src/logic.service.ts:237]: You missed these

File: src/logic.service.ts [hunk 5]
```diff
@@ -188,6 +187,77 @@ export const handleGetConditionsForEntity = async (params: GetEntityConditions):
...
+    xprtnDtTm = new Date().toISOString();
+  } else {
+    if (hasDateExpired(new Date(xprtnDtTm))) {
+      return { code: 400, message: 'Expiration time date provided was before the current time date.' };
+    }
+
+    if (!isDateValid(xprtnDtTm)) {
+      return { code: 400, message: 'Expiration time date provided was invalid.' };
+    }
+  }
```
Comment (Lenbkr) [inline src/logic.service.ts:204]: There is no need to check a non optional parameter as this the only scenario where xprtnDtTm will fail if it is will not parse as a date or is set to `''` which you can check for both in the `isDateValid` function. (Which should happen first)

```suggestion
  if(!isDateValid(xprtnDtTm)) {...} // covers what I highlighted above. Should also happen first
  const date = new Date(xprtnDtTm);
  if(hasDateExpired(date)) {...}// expired logic
  // normal execution  
```

**Review by Justus-at-Tazama — APPROVED** [review]

**Review by Lenbkr — APPROVED** [review]

**Review by Justus-at-Tazama — APPROVED** [review]

**Pull request merged**